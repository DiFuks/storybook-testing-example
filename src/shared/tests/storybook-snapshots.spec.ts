import type { Meta, StoryFn } from '@storybook/react';
import { composeStories } from '@storybook/react';
import { render } from '@testing-library/react';
import path from 'node:path';
import { describe, expect, test } from 'vitest';

type StoryFile = {
	default: Meta;
	[name: string]: StoryFn | Meta;
};

const compose = (entry: StoryFile): ReturnType<typeof composeStories<StoryFile>> => {
	try {
		return composeStories(entry);
	} catch (e) {
		throw new Error(`There was an issue composing stories for the module: ${JSON.stringify(entry)}, ${e}`);
	}
};

const getAllStoryFiles = () => {
	const storyFiles = Object.entries(
		import.meta.glob<StoryFile>(`../../../src/shared/ui/**/*.(stories|story).@(js|jsx|mjs|ts|tsx)`, {
			eager: true,
		}),
	);

	return storyFiles.map(([filePath, storyFile]) => {
		const storyDir = path.dirname(filePath);
		const componentName = path.basename(filePath).replace(/\.(stories|story)\.[^./]+$/, ``);

		return { filePath, storyFile, componentName, storyDir };
	});
};

describe(`Stories Snapshots`, () => {
	getAllStoryFiles().forEach(({ storyFile, componentName }) => {
		const meta = storyFile.default;
		const title = meta.title || componentName;

		describe(title, () => {
			const stories = Object.entries(compose(storyFile)).map(([name, story]) => ({ name, story }));

			if (stories.length <= 0) {
				throw new Error(
					`No stories found for this module: ${title}. Make sure there is at least one valid story for this module.`,
				);
			}

			stories.forEach(({ name, story }) => {
				test(name, async () => {
					const utils = render(story());
					await new Promise(resolve => {
						setTimeout(resolve, 1);
					});
					expect(utils.container).toMatchSnapshot();
				});
			});
		});
	});
});
