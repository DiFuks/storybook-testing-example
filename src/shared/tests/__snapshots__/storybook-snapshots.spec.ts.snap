// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Stories Snapshots > Button > Primary 1`] = `
.c0 {
  padding: 10px 20px;
  border: none;
  border-radius: 30px;
  color: white;
  background-color: #083e75;
  cursor: pointer;
  font-size: 16px;
  transition: background-color 0.2s;
}

.c0:hover {
  background-color: #083e75;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 0 2px rgba(0,123,255,0.25);
}

<div>
  <button
    class="c0"
    variant="primary"
  >
    Primary
  </button>
</div>
`;

exports[`Stories Snapshots > Button > Secondary 1`] = `
.c0 {
  padding: 10px 20px;
  border: none;
  border-radius: 30px;
  color: white;
  background-color: #6c757d;
  cursor: pointer;
  font-size: 16px;
  transition: background-color 0.2s;
}

.c0:hover {
  background-color: #545b62;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 0 2px rgba(0,123,255,0.25);
}

<div>
  <button
    class="c0"
    variant="secondary"
  >
    Secondary
  </button>
</div>
`;
